### PR TITLE
[JEP-222] Enable WebSocket mode by passing JENKINS_WEB_SOCKET=true

### PR DESCRIPTION
**Pull request by jglick**
See [JEP-222](https://jenkins.io/jep/222). Alternative to #129. Works with https://github.com/jenkinsci/remoting/pull/357. Requires #147.

### CODE REVIEW

**Comment by slide** [comment]
We will need to update the Windows side as well (https://github.com/jenkinsci/docker-jnlp-slave/blob/master/jenkins-agent.ps1)

**Comment by jglick** [comment]
Yes, I did not try that yet but I could.

**Comment by jglick** [comment]
> the upstream pull request in Remoting needs to be released first

Yes, this is on hold until the core changes are released.

File: jenkins-agent.ps1 [hunk 1]
```diff
@@ -67,6 +69,10 @@ if(![System.String]::IsNullOrWhiteSpace($Cmd)) {
 		$WorkDir = " -workDir `"$WorkDir`""
 	}
 
+	if("true".Equals($Env:JENKINS_WEB_SOCKET)) {
```
Comment (jglick) [inline jenkins-agent.ps1]: Tested a similar condition in a local PowerShell prompt, but did not try this script in context.

Comment (slide) [reply jenkins-agent.ps1]: You don't need to use .Equals in .NET like you do in Java. You can just do `$env:JENKINS_WEB_SOCKET -eq "true"`

This is more Powershell-ish

Comment (jglick) [reply jenkins-agent.ps1]: And seems to work with a null value too. Will edit accordingly.

**Comment by jglick** [comment]
> [Remoting 4.0] must be integrated and released in the parent image

Sure. I presume someone like @jeffret-b would be doing this as a routine task.

**Comment by jeffret-b** [comment]
I was hoping to get to that today, but may have more pressing tasks.

**Comment by jglick** [comment]
I think there is no rush. This change is needed for people who wish to use WebSocket transport for agents managed by the `kubernetes` plugin’s cloud, which is not the typical use case. Otherwise you could just rely on the `*.jnlp` parsing to set `-webSocket` (see #87). I suppose there are other people using this launcher script for Dockerized inbound agents who could benefit. And of course it only matters if you are running 2.217+.

**Comment by jeffret-b** [comment]
I managed to squeeze it in before lunch. The PRs are up.

**Comment by jglick** [comment]
So can you merge & release this @jeffret-b, or whoever is maintaining this repo?

**Comment by jeffret-b** [comment]
I don't have permission in this repository. Maybe @slide or @oleg-nenashev

**Comment by jglick** [comment]
@MarkEWaite thanks! Now can you can a release please? https://github.com/jenkinsci/docker-jnlp-slave/compare/3.40-1...master

**Comment by MarkEWaite** [comment]
I would like to leave this for @oleg-nenashev to release, since he's released the most recent several versions of it and I've never released a version.  If @oleg-nenashev has not been able to release it by tomorrow, I'll attempt the release.

**Comment by jglick** [comment]
ping @MarkEWaite @oleg-nenashev

**Comment by MarkEWaite** [comment]
Sorry that I missed this one @jglick.  As I look at the repository, the release process is not immediately clear to me.  I'll talk with @oleg-nenashev tomorrow and learn how to release from this repository.

**Comment by jglick** [comment]
I am not sure either. It may suffice to just create a tag and wait for Docker Hub automated builds to kick in.

**Comment by jglick** [comment]
IIUC from the `Jenkinsfile`, Linux publishing is passive, whereas a Jenkins project publishes the Windows version.

**Comment by slide** [comment]
The linux publishing should be done on dockerhub builds, though I am not sure how the tagging for specific versions takes place. For the Windows versions, we are publishing them via trusted.ci, we do not currently do a version specific release, I would need to look into that.

**Comment by oleg-nenashev** [comment]
on it

**Comment by MarkEWaite** [comment]
The [4.0.1-1 release](https://github.com/jenkinsci/docker-jnlp-slave/releases/tag/4.0.1-1) is now available and ready to use.  A [CONTRIBUTING](https://github.com/jenkinsci/docker-jnlp-slave/blob/master/CONTRIBUTING.md) document has also been added which includes a video link to a session that describes the release process.

**Comment by jglick** [comment]
```console
$ docker run --rm jenkinsci/jnlp-slave:4.0.1-1 2>&1 | fgrep -i websocket
 -webSocket                            : Make a WebSocket connection to Jenkins
```

:tada:

**Pull request merged**